spmd/distributed.py

```diff
@@ -0,0 +1,94 @@
+import io
+from dataclasses import replace
+from typing import Any
+
+import torch
+import torch.distributed.distributed_c10d as c10d
+from torch.distributed import ProcessGroup
+from torch.distributed._shard.checkpoint.default_planner import (
+    DefaultLoadPlanner,
+    DefaultSavePlanner,
+)
+from torch.distributed._shard.checkpoint.metadata import (
+    STATE_DICT_TYPE,
+    Metadata,
+    MetadataIndex,
+)
+from torch.distributed._shard.checkpoint.planner import ReadItem, SavePlan
+
+
+def get_ranks(pg: ProcessGroup):
+    """
+    Return an array of global ranks for a given process group.
+    """
+    return [c10d._get_global_rank(pg, i) for i in range(pg.size())]
+
+
+class ProcessGroupAwareSavePlanner(DefaultSavePlanner):
+    """
+    SavePlanner class defining the protocol used by save_state_dict() to plan
+    a process group aware save plan.
+    """
+
+    def create_local_plan(self) -> SavePlan:
```

Comment:
Could you also add some tests for these planner changes? You can create a `test_checkpoint.py` in the folder of `/test/spmd/`